File: peer-top20-report/data_processor.py
"""数据处理模块 - 移植自 sameIndustryAnalysis.jsx"""

from __future__ import annotations

import html
import re
from dataclasses import dataclass
from typing import Any


def parse_to_number(value: Any) -> float:
    if value is None or value == "":
        return float("nan")
    text = str(value).replace(",", "").replace("+", "").replace("$", "").replace(" ", "")
    try:
        return float(text)
    except ValueError:
        return float("nan")


def calculate_avg(values: list[float], fix_number: int = 2) -> str | int | float:
    valid = [v for v in values if v == v]  # filter NaN
    if not valid:
        return "N/A"
    total = sum(valid)
    if fix_number == -1:
        return int(total / len(valid))
    return round(total / len(valid), fix_number)


def unescape_html(text: str) -> str:
    if not text:
        return ""
    return html.unescape(re.sub(r"<[^>]+>", "", str(text)))


@dataclass
class PeerRow:
    rank: int
    company_name: str
    home: str
    main_products: str
    total_product_count: str
    page_views: str
    inquiries: str
    inquiry_rate: str
    transaction_number: str
    transaction_price: str
    display_star_level: str
    supplier_year: str


@dataclass
class CategoryGroup:
    category: str
    rows: list[PeerRow]
    summary: dict[str, Any]


@dataclass
class KeywordReport:
    keyword: str
    categories: list[CategoryGroup]
    default_category: str


def build_peer_row(record: dict[str, Any], rank: int) -> PeerRow:
    page_views = parse_to_number(record.get("pageViews"))
    inquiries = parse_to_number(record.get("iquiries"))
    if page_views == page_views and page_views > 0 and inquiries == inquiries:
        inquiry_rate = f"{(inquiries / page_views) * 100:.2f}%"
    else:
        inquiry_rate = "N/A"

    return PeerRow(
        rank=rank,
        company_name=str(record.get("companyName", "")),
        home=str(record.get("home", "")),
        main_products=unescape_html(record.get("mainProducts", "")),
        total_product_count=str(record.get("totalProductCount", "")),
        page_views=str(record.get("pageViews", "")),
        inquiries=str(record.get("iquiries", "")),
        inquiry_rate=inquiry_rate,
        transaction_number=str(record.get("transactionNumber", "")),
        transaction_price=str(record.get("transactionPrice", "")),
        display_star_level=str(record.get("displayStarLevel", "")),
        supplier_year=str(record.get("supplierYear", "")),
    )


def build_summary(records: list[dict[str, Any]]) -> dict[str, Any]:
    total_product_count = calculate_avg(
        [parse_to_number(r.get("totalProductCount")) for r in records], 0
    )
    page_views = calculate_avg([parse_to_number(r.get("pageViews")) for r in records], 0)
    inquiries = calculate_avg([parse_to_number(r.get("iquiries")) for r in records], 0)
    transaction_number = calculate_avg(
        [parse_to_number(r.get("transactionNumber")) for r in records], -1
    )
    transaction_price = calculate_avg(
        [parse_to_number(r.get("transactionPrice")) for r in records], 0
    )
    display_star_level = calculate_avg(
        [parse_to_number(r.get("displayStarLevel")) for r in records], -1
    )
    supplier_year = calculate_avg(
        [parse_to_number(r.get("supplierYear")) for r in records], -1
    )

    page_views_num = parse_to_number(page_views)
    inquiries_num = parse_to_number(inquiries)
    if page_views_num == page_views_num and page_views_num > 0 and inquiries_num == inquiries_num:
        inquiry_rate = f"{(inquiries_num / page_views_num) * 100:.2f}%"
    else:
        inquiry_rate = "N/A"

    return {
        "total_product_count": total_product_count,
        "page_views": page_views,
        "inquiries": inquiries,
        "inquiry_rate": inquiry_rate,
        "transaction_number": transaction_number,
        "transaction_price": f"${transaction_price}" if transaction_price != "N/A" else "N/A",
        "display_star_level": display_star_level,
        "supplier_year": supplier_year,
    }


def build_category_group(category: str, records: list[dict[str, Any]]) -> CategoryGroup:
    top20 = records[:20]
    rows = [build_peer_row(record, index + 1) for index, record in enumerate(top20)]
    return CategoryGroup(category=category, rows=rows, summary=build_summary(top20))


def process_same_industry_data(data: dict[str, Any]) -> list[KeywordReport]:
    """解析 sameIndustryAnalyseList 数据结构，生成报告所需内容。"""
    source = data.get("sameIndustryAnalyseList", data)
    if not isinstance(source, list):
        raise ValueError("输入数据缺少 sameIndustryAnalyseList 字段或格式不正确")

    reports: list[KeywordReport] = []
    for item in source:
        keyword = str(item.get("keyword", ""))
        grouped = item.get("effectDataCategoryGrouped") or []
        categories: list[CategoryGroup] = []

        if grouped:
            for group in grouped:
                category = str(group.get("category", "未分类"))
                values = group.get("value") or []
                categories.append(build_category_group(category, values))
            default_category = categories[0].category if categories else "未分类"
        else:
            effect_data = item.get("effectData") or []
            default_category = str(
                effect_data[0].get("category", "默认类目") if effect_data else "默认类目"
            )
            categories = [build_category_group(default_category, effect_data)]

        reports.append(
            KeywordReport(
                keyword=keyword,
                categories=categories,
                default_category=default_category,
            )
        )

    return reports

File: peer-top20-report/dataProcessor.js
/** 数据处理 - 移植自 sameIndustryAnalysis.jsx */

export function parseToNumber(value) {
  if (value === null || value === undefined || value === '') {
    return Number.NaN;
  }
  const text = String(value).replace(/[,+$ ]/g, '');
  const num = Number.parseFloat(text);
  return Number.isNaN(num) ? Number.NaN : num;
}

export function calculateAvg(values, fixNumber = 2) {
  const valid = values.filter((v) => !Number.isNaN(v));
  if (!valid.length) {
    return 'N/A';
  }
  const total = valid.reduce((sum, item) => sum + item, 0);
  if (fixNumber === -1) {
    return Math.trunc(total / valid.length);
  }
  return Number((total / valid.length).toFixed(fixNumber));
}

export function unescapeHtml(text) {
  if (!text) {
    return '';
  }
  return String(text)
    .replace(/<[^>]+>/g, '')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'");
}

function resolvePlatformCategory(record) {
  const direct = String(record.platformCategory || '').trim();
  if (direct && direct !== '-') {
    return unescapeHtml(direct);
  }

  const path = record.categoryPath;
  if (Array.isArray(path) && path.length) {
    const fromPath = path
      .map((item) => String(item?.categoryName || '').trim())
      .filter(Boolean)
      .join(' > ');
    if (fromPath) {
      return unescapeHtml(fromPath);
    }
  }

  const categoryName = String(record.categoryName || '').trim();
  if (categoryName) {
    return unescapeHtml(categoryName);
  }

  return '-';
}

function buildListedProductRow(product) {
  const pageViews = parseToNumber(product.pageViews);
  const inquiries = parseToNumber(product.iquiries);
  let inquiryRate = 'N/A';
  if (!Number.isNaN(pageViews) && pageViews > 0 && !Number.isNaN(inquiries)) {
    inquiryRate = `${((inquiries / pageViews) * 100).toFixed(2)}%`;
  }
  return {
    productId: String(product.productId || ''),
    mainProducts: unescapeHtml(product.mainProducts || ''),
    platformCategory: unescapeHtml(product.platformCategory || '-'),
    pageViews: String(product.pageViews || ''),
    inquiries: String(product.iquiries || ''),
    inquiryRate,
    productDetailUrl: String(product.productDetailUrl || ''),
  };
}

export function buildPeerRow(record, rank) {
  const pageViews = parseToNumber(record.pageViews);
  const inquiries = parseToNumber(record.iquiries);
  let inquiryRate = 'N/A';
  if (!Number.isNaN(pageViews) && pageViews > 0 && !Number.isNaN(inquiries)) {
    inquiryRate = `${((inquiries / pageViews) * 100).toFixed(2)}%`;
  }

  return {
    rank,
    companyName: String(record.companyName || ''),
    home: String(record.home || ''),
    mainProducts: unescapeHtml(record.mainProducts || ''),
    platformCategory: resolvePlatformCategory(record),
    pageViews: String(record.pageViews || ''),
    inquiries: String(record.iquiries || ''),
    inquiryRate,
    transactionNumber: String(record.transactionNumber || ''),
    transactionPrice: String(record.transactionPrice || ''),
    displayStarLevel: String(record.displayStarLevel || ''),
    supplierYear: String(record.supplierYear || ''),
    listedProductCount: Number(record.listedProductCount) || 1,
    listedProducts: Array.isArray(record.listedProducts)
      ? record.listedProducts.map(buildListedProductRow)
      : [],
  };
}

export function buildSummary(records) {
  const pageViews = calculateAvg(records.map((r) => parseToNumber(r.pageViews)), 0);
  const inquiries = calculateAvg(records.map((r) => parseToNumber(r.iquiries)), 0);
  const transactionNumber = calculateAvg(
    records.map((r) => parseToNumber(r.transactionNumber)),
    -1,
  );
  const transactionPrice = calculateAvg(
    records.map((r) => parseToNumber(r.transactionPrice)),
    0,
  );
  const displayStarLevel = calculateAvg(
    records.map((r) => parseToNumber(r.displayStarLevel)),
    -1,
  );
  const supplierYear = calculateAvg(records.map((r) => parseToNumber(r.supplierYear)), -1);

  const pageViewsNum = parseToNumber(pageViews);
  const inquiriesNum = parseToNumber(inquiries);
  let inquiryRate = 'N/A';
  if (!Number.isNaN(pageViewsNum) && pageViewsNum > 0 && !Number.isNaN(inquiriesNum)) {
    inquiryRate = `${((inquiriesNum / pageViewsNum) * 100).toFixed(2)}%`;
  }

  return {
    pageViews,
    inquiries,
    inquiryRate,
    transactionNumber,
    transactionPrice: transactionPrice === 'N/A' ? 'N/A' : `$${transactionPrice}`,
    displayStarLevel,
    supplierYear,
  };
}

export function buildCategoryGroup(category, records, { inquiryThreshold = 0 } = {}) {
  const filtered =
    inquiryThreshold > 0
      ? records.filter((record) => {
          const inquiries = parseToNumber(record.iquiries);
          return !Number.isNaN(inquiries) && inquiries > inquiryThreshold;
        })
      : records;
  const sorted = [...filtered].sort((a, b) => {
    const left = parseToNumber(a.iquiries);
    const right = parseToNumber(b.iquiries);
    if (Number.isNaN(left) && Number.isNaN(right)) return 0;
    if (Number.isNaN(left)) return 1;
    if (Number.isNaN(right)) return -1;
    return right - left;
  });
  const top20 = sorted.slice(0, 20);
  return {
    category,
    rows: sorted.map((record, index) => buildPeerRow(record, index + 1)),
    summary: buildSummary(top20),
    totalCount: sorted.length,
  };
}

function resolveInquiryThreshold(data, item) {
  const fromItem =
    item?.scrapingStats?.inquiryThreshold ??
    item?.timings?.inquiryThreshold ??
    item?.inquiryThreshold;
  const fromRoot =
    data?.inquiryThreshold ?? data?.timings?.inquiryThreshold ?? data?.scrapingStats?.inquiryThreshold;
  const value = fromItem ?? fromRoot ?? 0;
  const parsed = Number(value);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : 0;
}

export function processSameIndustryData(data) {
  const source = data.sameIndustryAnalyseList || data;
  if (!Array.isArray(source)) {
    throw new Error('输入数据缺少 sameIndustryAnalyseList 字段或格式不正确');
  }

  return source.map((item) => {
    const keyword = String(item.keyword || '');
    const inquiryThreshold = resolveInquiryThreshold(data, item);
    const grouped = item.effectDataCategoryGrouped || [];
    let categories = [];
    let defaultCategory = '未分类';

    if (grouped.length) {
      categories = grouped
        .map((group) =>
          buildCategoryGroup(String(group.category || '未分类'), group.value || [], {
            inquiryThreshold,
          }),
        )
        .filter((category) => category.rows.length > 0);
      defaultCategory = categories[0]?.category || '未分类';
    } else {
      const effectData = item.effectData || [];
      defaultCategory = String(effectData[0]?.category || '默认类目');
      categories = [
        buildCategoryGroup(defaultCategory, effectData, { inquiryThreshold }),
      ].filter((category) => category.rows.length > 0);
    }

    return { keyword, categories, defaultCategory };
  });
}

export function countReportRows(reports) {
  return reports.reduce(
    (total, report) =>
      total + report.categories.reduce((sum, category) => sum + category.rows.length, 0),
    0,
  );
}

export function assertReportHasRows(reports) {
  if (!reports.length) {
    throw new Error('未找到可处理的同行数据');
  }
  if (countReportRows(reports) === 0) {
    throw new Error(
      '报告里没有同行店铺数据。请确认已登录阿里巴巴国际站、关键词有效且未被验证码拦截。',
    );
  }
}
